### DISCUSSION TITLE
Improved CodeSpaces Insights / Reports

### DISCUSSION THREAD

**Discussion by numbsafari** [community/community#9611] [Codespaces] [answered]
I apologize if these features exist, but I've been looking and so far cannot find them.

Basically, it would be really helpful to be able to track codespaces utilization across an org by user (and perhaps by repo as well). It would be helpful to know which users are consuming the most, etc., in case there is something that is contributing to high costs.

It'd be great to understand the number of active instances over time (e.g., do we mysteriously have spaces running at odd times when they shouldn't be?).

Right now, we can pretty much get a breakdown of cost by machine type for the current month.

I think making this available via either the org or repo "Insights" page would be really helpful (perhaps optionally) so that a team can have visibility on their resource consumption, rather than locking that behind the billing account info.

**Comment by tanmayeekamath** [comment, marked as answer]
Hey @numbsafari! Thanks for your feedback. 
Currently, you can find the breakdown of Codespaces usage for your org by going to Org settings -> Billing & plans -> Click on the Get usage report button. Once you complete the steps, you should receive the report via email. 
<img width="253" alt="image" src="https://user-images.githubusercontent.com/30297258/148109000-956309d6-23d2-41d6-8871-57cd38d9dfbe.png">

This should give you a breakdown of Codespaces costs by repo, SKU name and user. Would love to know if this breakdown is helpful? 

Having said that, there are a couple things I captured: 
- We need to make this functionality more visible & accessible perhaps by adding a section for it within the Codespaces settings tab for the org and repo. As an immediate update, we'll make sure to update our docs with this information to make it easy to find. 
- Add data around the state of Codespaces (active, suspended) across repos for the org. On similar lines, we'll soon be adding support for a retention policy on Codespaces that can be set by an org admin to help minimize wasteful resources arising from Codespaces that were created a long time ago and never used for a long time. Would this help to address your scenario? 

Thanks again for the inputs, looking forward to learn more around your scenario & needs based on above :) 

-- Tanmayee

**Reply by numbsafari** [reply]
Had a chance to play around with this last week.

Definitely helpful to at least have access to the data.

Having said that, I think having some kind of baseline visualizations in the insights page as noted above would be a lot more user friendly.

The retention policy you mentioned would be great. If there were a way to have a soft warning to let a user know that certain spaces will soon be deleted, that would be awesome. e.g., set a policy that after X days of inactivity, the user will see a warning (similar to dependabot?) that unused spaces will be removed in, say, another 72 hours. Then, after 10 days of inactivity, it gets deleted.

It looks like my other concerns are addressed in the overall security audit logs.

Thanks!
